src/main/java/kohn/maze/Maze.java

```diff
@@ -1,368 +1,179 @@
+        int y = 0;
+        Cell cell = maze[x][y];
+
+        while (!containsUnvisitedCells()) {
+            if (removeVisitedNeighbors(cell) != 0) {
+                Cell neighbor = getRandomNeighbor(cell);
+                if(neighbor != null) {
+                    removeWalls(cell, neighbor);
+                    stack.push(cell);
+                    cell = neighbor;
+                    cell.visited = true;
+                }
             }
-            if ((theMaze[col - 3][row].equals("#") || theMaze[col - 1][row].equals("#"))
-                    || (theMaze[col - 2][row - 1].equals("#")  || theMaze[col - 2][row + 1].equals("#"))){
-
-                return checkNeighbors(theMaze, current, directions);
+            else {
+                cell = stack.pop();
             }
         }
-        else if (check.equals("E")) {
-            if (current.getRow() + 1 >= size) {
-
-                return checkNeighbors(theMaze, current, directions);
-            }
-            if (((theMaze[col + 1][row + 2].equals("#") || theMaze[col - 1][row + 2].equals("#"))
-                    || (theMaze[col][row + 1].equals( "#") || theMaze[col][row + 3].equals("#")))) {
+    }
 
-                return checkNeighbors(theMaze, current, directions);
-            }
+    private void assignNeighbors(Cell cell) {
+        if (cell.getY() != 0) { //assigns left neighbor
+            cell.neighbors.add(maze[cell.getX()][cell.getY() - 1]);
         }
-        else if (check.equals("S")) {
-            if (current.getCol() + 1 >= size) {
-
-                return checkNeighbors(theMaze, current, directions);
-            }
-            if (((theMaze[col + 1][row].equals("#") || theMaze[col + 3][row].equals("#"))
-                    || (theMaze[col + 2][row - 1].equals("#") || theMaze[col + 2][row + 1].equals("#")))) {
 
-                return checkNeighbors(theMaze, current, directions);
-            }
+        if (cell.getX() != 0) { //assigns top row
+            cell.neighbors.add(maze[cell.getX() - 1][cell.getY()]);
         }
-        else if (check.equals("W")) {
-            if (current.getRow() - 1 < 0) {
 
-                return checkNeighbors(theMaze, current, directions);
-            }
-            if (((theMaze[col - 1][row - 2].equals("#") || theMaze[col + 1][row - 2].equals("#"))
-                    || (theMaze[col][row - 3].equals("#") || theMaze[col][row - 1].equals("#")))) {
+        if (cell.getX() != rows - 1) { //assigns south cell
+            cell.neighbors.add(maze[cell.getX() + 1][cell.getY()]);
+        }
 
-                return checkNeighbors(theMaze, current, directions);
-            }
+        if (cell.getY() != cols - 1) {//east cell
+            cell.neighbors.add(maze[cell.getX()][cell.getY() + 1]);
         }
-        return check;
     }
 
+    private Cell getRandomNeighbor(Cell cell) {
+        if (cell.neighbors.size() == 0) {
+            return null;
+        }
+        Random random = new Random();
```

Comment:
create `random` as a constant, do not recreate it each time you need a random number